### ISSUE TITLE
Re-run CESAR jobs crash- "some para/nextflow processes died"

### ISSUE THREAD

**Issue by pfb2201** [hillerlab/TOGA#146]
Hello! First of all, thank you so much for creating this amazing tool! I am trying to run the TOGA pipeline to annotate and infer orthologues from a group of butterfly genomes from the same family. I am running this on my university's cluster which uses slurm. I was able to successfully run both the micro and final tests, and run the entire TOGA pipeline on one of my species (after editing the CESAR job's config file to tell it to use the higher memory partition on our cluster). 

I am having trouble with the CESAR portion of the pipeline-- for the rest of my species, there are always a few jobs that initially get assigned to a lower memory bucket, fail, and then get re-run at the next memory bucket. These re-run jobs always fail, but I am having a hard time figuring out what exactly is the problem (do they need more memory? Should I optimize the cesar buckets? Is it a quirk with nextflow?) from the logs. The logs from the re-run CESAR jobs show that they executed, but then the pipeline doesn't continue. The only indication of a problem is the message "some para/nextflow processes died" that gets printed onto the terminal console (this message doesn't show up on the TOGA run log).

I have attached the script I used, and the entire TOGA run log and nextflow logs for the re-run CESAR jobs. Any help to figure out where to look for the problem would be very much appreciated! I would also like to know how to run the problematic CESAR jobs individually to figure out the exact problem that it's having, since I cannot pinpoint the cause of the error from the full pipeline.

Info on the installation: I created a conda environment (python 3.10) and downloaded all the dependencies using conda (twobitreader=3.1.7 pandas==2.1.1 numpy==1.26.0 scikit-learn==1.3.1 joblib==1.3.2 h5py==3.10.0) and pip (xgboost==2.0.1). My cluster had nextflow (23.0.4) and openjdk (17.0.2) installed already as modules. I then git cloned the TOGA repository and followed the rest of the installation/testing instructions as shown in the Github. 

[run_toga.sh.zip](https://github.com/hillerlab/TOGA/files/14270048/run_toga.sh.zip)(https://github.com/hillerlab/TOGA/files/14270050/cesar_jobs__RERUN_TOGA_Pieris_r_0212_at_5295389_q_500.log.zip)

**Comment by pfb2201** [comment]
Thank you for the reply! I don't know if they die from lack of memory, but the TOGA run log states the following when CESAR jobs crash:
`Checking whether all CESAR results are complete
1 CESAR jobs crashed, trying to run again...
!!RERUN CESAR JOBS: Pushing 4 jobs into 500 GB queue
Selected parallelization strategy: nextflow
Parallel manager: pushing job nextflow /gpfs/data/kronforst-lab/pfb1/conda/envs/TOGA/TOGA/execute_joblist.nf --joblist /gpfs/data/kronforst-lab/pfb1/conda/envs/TOGA/TOGA/TOGA_Pieris_r_0212/_cesar_rerun_batch_500 -c /gpfs/data/kronforst-lab/pfb1/conda/envs/TOGA/TOGA/TOGA_Pieris_r_0212/temp/cesar_config_500_queue.nf`

these re-run jobs crash again. I am assuming it is from lack of memory, but am having a hard time tracking down the problem.

**Comment by MichaelHiller** [comment]
This is strange. We need ~300 GB to align the entire Titin gene. Or genes that span > 1Mb in the genome (big introns), but I think such genes will not exist in butterflies. Maybe memory is not the issue. 

Could you run a few of these jobs manually on a compute or the head node? And check before how large the query sequence is. And maybe watch with top from another terminal how much mem this job is really consuming.

**Comment by pfb2201** [comment]
Hello Dr. Hiller! Apologies for the delay-- our cluster had some routine maintenance done to it and was inaccessible. It seems that CESAR is not the problem-- the jobs that needed to be re-run work completely fine. They all had outputted results and used less than 100G of memory (20-80G according to "seff JOBID"). 

I looked more closely at the script that produced the initial error message "Error! Some para/nextflow processes died!" (parallel_jobs_manager_helpers.py), and that script gives the error when the bolded condition is met

```
    **if any(jm.return_code != 0 for jm in jobs_managers) and die_if_sc_1 is True:**
        # some para/nextflow job died: critical issue
        # if die_if_sc_1 is True: terminate the program
        err = "Error! Some para/nextflow processes died!"
        # TODO: think about the best error class
        raise AssertionError(err)
```
Given that the jobs ran (and should result in return_code=0), the error must be related to "die_if_sc_1=True" condition being met. I am not sure what this condition means exactly- it shows up once in the toga.py script. Any insight on this would be very helpful!

**Comment by MichaelHiller** [comment]
I am pretty sure the "die_if_sc_1=True" is just a flag telling TOGA to error exit if the return status is not 0. 
You can probably disable that (should be fine for this run where you know they all worked), but then you won't catch jobs that died in another run, giving you incomplete results. 

The key question is why do the jobs run when you run them manually but die when you submit them to the cluster queue. 
Can you try running them manually on a node where they crashed? Maybe something is not properly installed at this or all compute nodes? 
Or when running the job via Slurm, some configuration is not properly set. 

@kirilenkobm How can one get a detailed log of the cluster job to figure out at which step it crashes?

**Comment by pfb2201** [comment]
Hello Dr. Hiller! I don't think that it is a slurm issue because my other run (with the query being a different species but the reference input data is all the same) ran without issues. I re-ran the entire TOGA pipeline (without deactivating the die_if_sc_1=TRUE line) with and without the --ms flag (and both with --kt to better trace issues). The pipeline runs completely when I include the --ms flag, but my understanding with this condition is that one should only include it if the reference annotation contains premature stop codons. In my initial run with the species that worked, I did not include the --ms flag and it ran without issues. I therefore should not include this flag, correct? Given that it seems that the annotation file of my reference is not the problem. Of course, maybe there are some genes that have premature stop codons that do not align with the species that I was able to run TOGA with successfully. Is there any drawback to using this flag if I am not sure if my reference contains in-frame stop codons?

Additionally, I found that the re-run CESAR jobs on a TOGA run without the --ms flag (which fails saying "some para/nextflow processes died") DO have results that are printed as files (in temp/cesar_results). These are printed in the same directory as the rest of the CESAR job results. But for some reason, the TOGA pipeline doesn't continue with merging results even though all results are in the same directory. 

For now, I will disable the condition that stops the pipeline with my current species and will see if this is an issue with the rest of my butterfly genomes.

**Comment by MichaelHiller** [comment]
--ms is masking stop codons that could be selenocysteine TGAs in the reference. 
If the reference species and reference annotation is the same and it works well without --ms for one query, then it is weird that it crashes for another query. But you could be right that the affected genes may not align in other queries and therefore are not aligned by CESAR, which should then throw an error if there is an unmasked internal stop. 
I am glad to hear that it runs through if you set --ms

There is no real downside to masking such stops. The alternative is removing the respective genes from your reference annotation, but that likely leads to missing genes. 

Wrt NextFlow and re-running some jobs, I hope @kirilenkobm can have a look.

**Comment by kirilenkobm** [comment]
Michael, thank you a lot for your help in this matter!

Hi @pfb2201 !

> I am having trouble with the CESAR portion of the pipeline-- for the rest of my species, there are always a few jobs that initially get assigned to a lower memory bucket, fail, and then get re-run at the next memory bucket. These re-run jobs always fail, but I am having a hard time figuring out what exactly is the problem (do they need more memory? Should I optimize the cesar buckets? Is it a quirk with nextflow?) from the logs. 

Increasing memory requirements is usually a default measure, but memory is often not the root cause.

The problematic CESAR jobs should be listed in cesar_jobs_crashed.txt. A practical approach to diagnose the issue is to copy and paste a failed command to run it directly. I acknowledge the logging of individual jobs could be improved, and I apologize for any inconvenience this may cause.

For various reasons, ranging from incorrect reference annotation to significant memory demands, some transcripts indeed cannot be processed. It seems that in your case, the presence of stop codons in the reference might have caused the CESAR jobs to fail. I'm considering setting the --ms parameter as default.

**Comment by pfb2201** [comment]
Thanks so much! I really appreciate the help. Adding the --ms flag got rid of the pipeline dying problem. I'm gonna go ahead and close the comment!
